std/algorithm/searching.d

```diff
@@ -297,7 +297,7 @@ private:
     ptrdiff_t[ElementType!(Range)] occ;         // GC allocated
     Range needle;
 
-    ptrdiff_t occurrence(ElementType!(Range) c)
+    ptrdiff_t occurrence(ElementType!(Range) c) scope
     {
         auto p = c in occ;
```

Comment:
We should better leave this to inference as `opHash` or `opEquals` of the element type could possibly escape a reference to one of the keys of the `occ` hash table. Somewhat unlikely but still possible, e.g. some cache of expensive hashing.

Reply:
This sounds like rejection of PR, thus I will close it

Reply:
@MartinNowak   PR reopened:
I didn't gratuitously add scope, because for the implicit this parameter, it doesn't get inferred currently, as this will show. Temporarily apply these changes:
```
diff --git a/std/algorithm/searching.d b/std/algorithm/searching.d
index 42e128f..084bfc0 100644
--- a/std/algorithm/searching.d
+++ b/std/algorithm/searching.d
@@ -360,7 +360,7 @@ public:
     }
 
     ///
-    Range beFound(Range haystack)
+    Range beFound(Range haystack) @safe
     {
         import std.algorithm.comparison : max;
 
@@ -1839,7 +1839,7 @@ if (isInputRange!InputRange)
 }
 
 /// ditto
-R1 find(alias pred = "a == b", R1, R2)(R1 haystack, scope R2 needle)
+R1 find(alias pred = "a == b", R1, R2)(R1 haystack, scope R2 needle) @safe
 if (isForwardRange!R1 && isForwardRange!R2
         && is(typeof(binaryFun!pred(haystack.front, needle.front)) : bool))
 {
@@ -2408,7 +2408,7 @@ if (Ranges.length > 1 && is(typeof(startsWith!pred(haystack, needles))))
  * such position exists, returns `haystack` advanced to termination).
  */
 RandomAccessRange find(RandomAccessRange, alias pred, InputRange)(
-    RandomAccessRange haystack, scope BoyerMooreFinder!(pred, InputRange) needle)
+    RandomAccessRange haystack, scope BoyerMooreFinder!(pred, InputRange) needle) @safe
 {
     return needle.beFound(haystack);
 }
diff --git a/std/range/package.d b/std/range/package.d
index b179f98..e1609cc 100644
--- a/std/range/package.d
+++ b/std/range/package.d
@@ -9715,7 +9715,7 @@ if (isInputRange!Range)
     {
         return predFun(rhs, lhs);
     }
-    private Range _input;
+    /*private*/ Range _input;
 
     // Undocummented because a clearer way to invoke is by calling
     // assumeSorted.
```
make -f posix.mak std/algorithm/searching.test (aa[std.algorithm.searching]=-dip1000 in #6278)
results in:
...
```
T=`mktemp -d /tmp/.dmd-run-test.XXXXXX` &&                                                              \
  (                                                                                                     \
    ../dmd/generated/linux/release/64/dmd -od$T -conf= -I../druntime/import  -w -de -dip25 -m64 -fPIC -transition=complex -O -release -dip1000  -main -unittest generated/linux/release/64/libphobos2.a -defaultlib= -debuglib= -L-ldl -cov -run std/algorithm/searching.d ;     \
    RET=$? ; rm -rf $T ; exit $RET                                                                   \
  )
std/algorithm/searching.d(1969): Error: scope variable needle assigned to non-scope parameter r2 calling std.algorithm.comparison.mismatch!("a == b", SortedRange!(int[], "a < b"), SortedRange!(int[], "a < b")).mismatch
std/algorithm/searching.d(2101): Error: template instance `std.algorithm.searching.find!("a == b", SortedRange!(int[], "a < b"), SortedRange!(int[], "a < b"))` error instantiating
std/algorithm/searching.d(2413): Error: scope variable needle assigned to non-scope parameter this calling std.algorithm.searching.BoyerMooreFinder!(binaryFun, string).BoyerMooreFinder.beFound
std/algorithm/searching.d(2424): Error: template instance `std.algorithm.searching.find!(string, binaryFun, string)` error instantiating
std/algorithm/searching.d(2413): Error: scope variable needle assigned to non-scope parameter this calling std.algorithm.searching.BoyerMooreFinder!(binaryFun, int[]).BoyerMooreFinder.beFound
std/algorithm/searching.d(2436): Error: template instance `std.algorithm.searching.find!(int[], binaryFun, int[])` error instantiating
posix.mak:571: die Regel für Ziel „std/algorithm/searching.test“ scheiterte (the rule for target ... failed)
```
Thus, BoyerMooreFinder.beFound doesn't get its this parameter inferred scope.
And if You extend changes above to
Range beFound(Range haystack) scope @safe
the same is true (not inferred) for   ptrdiff_t occurrence(ElementType!(Range) c):
```
std/algorithm/searching.d(379): Error: scope variable this assigned to non-scope parameter this calling std.algorithm.searching.BoyerMooreFinder!(binaryFun, string).BoyerMooreFinder.occurrence
std/algorithm/searching.d(1969): Error: scope variable needle assigned to non-scope parameter r2 calling std.algorithm.comparison.mismatch!("a == b", SortedRange!(int[], "a < b"), SortedRange!(int[], "a < b")).mismatch
std/algorithm/searching.d(2101): Error: template instance `std.algorithm.searching.find!("a == b", SortedRange!(int[], "a < b"), SortedRange!(int[], "a < b"))` error instantiating
std/algorithm/searching.d(379): Error: scope variable this assigned to non-scope parameter this calling std.algorithm.searching.BoyerMooreFinder!(binaryFun, int[]).BoyerMooreFinder.occurrence
std/algorithm/searching.d(2436): Error: template instance `std.algorithm.searching.find!(int[], binaryFun, int[])` error instantiating
posix.mak:571: die Regel für Ziel „std/algorithm/searching.test“ scheiterte (the rule for target ... failed)
```
I'm about to push an update for this PR

Reply:
Just a spontaneous idea: couldn't inout be modified to also infer scope?

Reply:
> Just a spontaneous idea: couldn't inout be modified to also infer scope?

No, inout converts implicitly to const, it could be legally escaped. It also could be new data from the heap if returned.